test/controllers/auth_controller_test.rb

```diff
@@ -19,15 +23,37 @@ class HealthCardsControllerTest < ActionDispatch::IntegrationTest
     get(auth_code_path,
         params: { client_id: Rails.application.config.client_id, redirect_uri: redirect_uri, state: state })
     assert_redirected_to "#{redirect_uri}?code=#{Rails.application.config.auth_code}&state=#{state}"
+    assert_no_cache_headers
   end
 
-  test 'token with incorrect code' do
-    post(auth_token_path, params: { code: 'code' })
-    assert_response :unauthorized
+  test 'token with incorrect parameter should return 400 invalid_request' do
+    post(auth_token_path)
+    assert_response :bad_request
+    assert_no_cache_headers
+    assert_equal('{"error":"invalid_request"}', @response.body)
+  end
+
+  test 'token with incorrect code should return 400 invalid_client' do
+    post(auth_token_path, params: { code: 'bad_code' })
+    assert_response :bad_request
+    assert_no_cache_headers
+    assert_equal('{"error":"invalid_client"}', @response.body)
   end
 
   test 'token with correct code' do
     post(auth_token_path, params: { code: Rails.application.config.auth_code })
     assert_response :success
+    assert_no_cache_headers
```

Comment:
Should we make a separate test that calls `assert_no_cache_headers`? Right now if something is wrong with the headers, every test will fail, which may make it slightly hard to debug from a unit test perspective. I don't think there is benefit to adding the two header asserts to every single test.

Reply:
I considered that, but then I would either make one test that tests all action methods for the correct headers (which then makes it ambiguous which method caused the error) or make individual tests for each method just to test the header (which seems redundant). Should I do the latter anyways?

Reply:
I'm with Mick on this one. It's only two controller actions, so one test for each action checking the cache headers rather than checking them in every test.

Reply:
In the case of "one test that tests all action methods for the correct headers," wouldn't an error in the correct headers be the fault of something within the `assert_no_cache_headers` function, and as a result propagate to _all_ of the action methods? If that is the case, I don't think there would be a situation where you would need to decipher which action method caused the errors, and you could just throw all the actions into one test.

If not, I think it might be worth being redundant to modularize the test cases, but it's probably not too big of a deal.

Reply:
The fact that both of the requests get their cache headers are set in the same way is an implementation detail that the tests don't need to know about. Also, having a separate test for each endpoint makes it more likely that we'll remember to implement that test if we add another action to the controller (such as a token refresh).

I would not modularize the tests or do anything fancy at all with them. The simpler a test is the better.

Reply:
I fixed this now, doing a separate test for the headers.